presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java

```diff
@@ -0,0 +1,137 @@
+    private final double max;
+    private final double[] weights;
+
+    public FixedNumericHistogram(int numBuckets, double min, double max)
+    {
+        checkArgument(numBuckets >= 2, "numBuckets must be >= 2");
+        checkArgument(min < max, "max must be larger than min");
+
+        this.numBuckets = numBuckets;
+        this.min = min;
+        this.max = max;
+        this.weights = new double[numBuckets];
+    }
+
+    public FixedNumericHistogram(SliceInput input)
+    {
+        checkArgument(input.readByte() == FORMAT_TAG, "Unsupported format tag");
+
+        numBuckets = input.readInt();
+        checkArgument(numBuckets >= 2, "numBuckets must be >= 2");
+        min = input.readDouble();
+        max = input.readDouble();
+        checkArgument(min < max, "max must be greater than min");
+        this.weights = new double[numBuckets];
+        input.readBytes(Slices.wrappedDoubleArray(weights), numBuckets * SizeOf.SIZE_OF_DOUBLE);
+    }
+
+    public int getRequiredBytesForSerialization()
+    {
+        return SizeOf.SIZE_OF_BYTE + // format
+                SizeOf.SIZE_OF_INT + // num buckets
+                SizeOf.SIZE_OF_DOUBLE + // min
+                SizeOf.SIZE_OF_DOUBLE + // max
+                SizeOf.SIZE_OF_DOUBLE * numBuckets; // weights
+    }
+
+    public void serialize(SliceOutput out)
+    {
+        out.appendByte(FORMAT_TAG)
+            .appendInt(numBuckets)
+            .appendDouble(min)
+            .appendDouble(max)
+            .appendBytes(Slices.wrappedDoubleArray(weights, 0, numBuckets));
+    }
+
+    public long estimatedInMemorySize()
+    {
+        return INSTANCE_SIZE + SizeOf.sizeOf(weights);
+    }
+
+    public void set(double value, double weight)
+    {
+        checkArgument(value >= min, "value must not be smaller than min");
+        checkArgument(value <= max, "value must not be larger than max");
+        final int ind = Math.min(
+                (int) (numBuckets * (value - min) / (max - min)),
+                numBuckets - 1);
+        weights[ind] = weight;
+    }
+
+    public void add(double value)
+    {
+        add(value, 1);
+    }
+
+    public void add(double value, double weight)
+    {
+        checkArgument(value >= min, "value must not be smaller than min");
+        checkArgument(value <= max, "value must not be larger than max");
+        final int ind = Math.min(
+                (int) (numBuckets * (value - min) / (max - min)),
+                numBuckets - 1);
+        weights[ind] += weight;
+    }
+
+    public void mergeWith(FixedNumericHistogram other)
+    {
+        checkArgument(numBuckets == other.numBuckets, "numBuckets must be equal");
+        checkArgument(min == other.min, "min must be equal");
+        checkArgument(max == other.max, "max must be equal");
+
+        for (int i = 0; i < numBuckets; ++i) {
+            weights[i] += other.weights[i];
+        }
+    }
+
+    public SortedMap<Double, Double> getBuckets()
```

Comment:
I would consider having an object that represents a bin, and returning an `Iterable<Bin>` here.

We don't need to change this implementation to internally use Bin objects.  The flat memory we are using is already great.  We can instead fake it by mutating a singleton `Bin` and mutate it with the next bin's numbers when `iterator#next` is called.  This way we avoid all of the autoboxing and the need to create a map, since all we want to do is iterate over the bins in ascending order.